# 🛠️ Active Learning Workshop: Implementing an Inverted Matrix (Jupyter + GitHub Edition)
## 🔍 Workshop Theme
*Readable, correct, and collaboratively reviewed code—just like in the real world.*


Welcome to the 90-minute workshop! In this hands-on session, your team will build an **Inverted Index** pipeline, the foundation of many intelligent systems that need fast and relevant access to text data — such as AI agents.

### 👥 Team Guidelines
- Work in teams of 3.
- Submit one completed Jupyter Notebook per team.
- The final notebook must contain **Markdown explanations** and **Python code**.
- Push your notebook to GitHub and share the `.git` link before class ends.

---
## 🔧 Workshop Tasks Overview

1. **Document Collection**
2. **Tokenizer Implementation**
3. **Normalization Pipeline (Stemming, Stop Words, etc.)**
4. **Build and Query the Inverted Index**

> Each step includes a sample **talking point**. Your team must add your own custom **Markdown + code cells** with a **second talking point**, and test your Inverted Index with **2 phrase queries**.




## 🧠 Learning Objectives
- Implement an **Inverted Matrix** using real-world data during the NLP process.
- Build **Jupyter Notebooks** with well-structured code and clear Markdown documentation.
- Use **Git and GitHub** for collaborative version control and code sharing.
- Identify and articulate coding issues ("**talking points**") and insert them directly into peer notebooks.
- Practice **collaborative debugging**, professional peer feedback, and improve code quality.

## 🧩 Workshop Structure (90 Minutes)
1. **Instructor Use Case Introduction** *(15 min)* – Set up teams of 3 people. Read and understand the workshop, plus submission instructions. Seek assistance if needed.
2. **Team Jupyter Notebook Development** *(45 min)* – Manual IR and Inverted Matrix coding + Markdown documentation (work as teams)
3. **Push to GitHub** *(15 min)* – Teams commit and push initial notebooks. **Make sure to include your names so it is easy to identify the team that developed the Min-Max code**.
4. **Instructor Review** - The instructor will go around, take notes, and provide coaching as needed, during the **Peer Review Round**
5. **Email Delivery** *(15 min)* – Each team send the instructor an email **with the *.git link** to the GitHub repo **(one email/team)**. Subject on the email is: PROG8245 - Inverted Matrix  Workshop, Team #_____.
6. **Push to the course GitHub** *(automated)* – Find the last code cell in this notebook, update your team number, and run the cell. It will push your notebook to the repo at `submissions/team#`

## 💻 Submission Checklist
- ✅ `IR_InvertedMatrix_Workshop.ipynb` with:
  - Demo code: Document Collection, Tokenizer, Normalization Pipeline, and Inverted Index.
  - Markdown explanations for each major step
  - **Labeled talking point(s)** and 2 phrase query tests
- ✅ `README.md` with:
  - Dataset description
  - Team member names
  - Link to the dataset and license (if public)
- ✅ GitHub Repo:
  - Public repo named `IR-invertedmatrix-workshop`
  - This is a group effort, so **choose one member of the team** to publish the repo
  - At least **one commit containing one meaningful talking point**

## 📄 Step 1: Document Collection


### 🗣 Instructor Talking Point:
> We begin by gathering a text corpus. To build a robust index, your vocabulary should include **over 2000 unique words**. You can use scraped articles, academic papers, or open datasets.

### 🔧 Your Task:
- Collect at least 20+ text documents.
- Ensure the vocabulary exceeds 2000 unique words.
- Load the documents into a list for processing.


In [19]:
import os
import requests
import feedparser

# Example: Download blog posts from a public RSS feed and save as .txt files
def fetch_and_save_blog_posts(rss_url, save_folder, max_posts=20):
    feed = feedparser.parse(rss_url)
    if not os.path.exists(save_folder):
        os.makedirs(save_folder)
    count = 0
    for entry in feed.entries:
        if count >= max_posts:
            break
        title = entry.get('title', 'untitled')
        content = entry.get('summary', '')
        # Clean filename
        filename = f"blog_{count+1}.txt"
        filepath = os.path.join(save_folder, filename)
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(f"{title}\n\n{content}")
        count += 1
    print(f"Saved {count} blog posts to {save_folder}")

# Example RSS feed: Python Software Foundation blog
rss_url = 'https://pyfound.blogspot.com/feeds/posts/default?alt=rss'
fetch_and_save_blog_posts(rss_url, 'sample_docs', max_posts=20)

Saved 20 blog posts to sample_docs


In [20]:
# Example: Load text files from a folder
import os

def load_documents(folder_path):
    documents = []
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):
            with open(os.path.join(folder_path, filename), 'r', encoding='utf-8') as file:
                documents.append(file.read())
    return documents

# Replace 'sample_docs/' with your actual folder
documents = load_documents('sample_docs/')
print(f"Loaded {len(documents)} documents.")


Loaded 20 documents.


## ✂️ Step 2: Tokenizer


### 🗣 Instructor Talking Point:
> The tokenizer breaks raw text into a stream of words (tokens). This is the foundation for every later step in IR and NLP.

### 🔧 Your Task:
- Implement a basic tokenizer that splits text into lowercase words.
- Handle punctuation removal and basic non-alphanumeric filtering.


In [21]:
import re

def tokenize(text):
    tokens = re.findall(r'\b\w+\b', text.lower())
    return tokens

# Test on one document
tokens = tokenize(documents[0])
print(tokens[:20])  # Preview first 20 tokens


['notice', 'of', 'python', 'software', 'foundation', 'bylaws', 'change', 'effective', 'july', '23', '2025', 'p', 'this', 'post', 'serves', 'as', 'notice', 'that', 'the', 'board']


## 🔁 Step 3: Normalization Pipeline (Stemming, Stop Word Removal, etc.)


### 🗣 Instructor Talking Point:
> Now we normalize tokens: convert to lowercase, remove stop words, apply stemming or affix stripping. This reduces redundancy and enhances search accuracy.

### 🔧 Your Task:
- Use `nltk` to remove stopwords and apply stemming.


In [22]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

nltk.download('stopwords')
stop_words = set(stopwords.words('english'))
stemmer = PorterStemmer()

def normalize_tokens(tokens):
    return [stemmer.stem(t) for t in tokens if t not in stop_words]

# Example: normalize one document
norm_tokens = normalize_tokens(tokens)
print(norm_tokens[:20])


['notic', 'python', 'softwar', 'foundat', 'bylaw', 'chang', 'effect', 'juli', '23', '2025', 'p', 'post', 'serv', 'notic', 'board', 'python', 'softwar', 'foundat', 'resolv', 'amend']


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Eespinosa\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


## 🔍 Step 4: Inverted Index


### 🗣 Instructor Talking Point:
> We now map each normalized token to the list of document IDs in which it appears. This is the core structure that allows fast Boolean and phrase queries.

### 🔧 Your Task:
- Build the inverted index using a dictionary.
- Add code to support phrase queries using positional indexing.


In [ ]:
from collections import defaultdict

def build_inverted_index(documents):
    index = defaultdict(list)
    for doc_id, text in enumerate(documents):
        tokens = normalize_tokens(tokenize(text))
        seen = set()
        for token in tokens:
            if token not in seen:
                index[token].append(doc_id)
                seen.add(token)
    return dict(index)

inverted_index = build_inverted_index(documents)
print(dict(list(inverted_index.items())[:10]))  # Preview first 10 terms


{'notic': [0], 'python': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19], 'softwar': [0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], 'foundat': [0, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14, 15], 'bylaw': [0, 14], 'chang': [0, 1, 2, 3, 5, 6, 7, 10, 12, 14, 15, 16, 17, 18, 19], 'effect': [0, 10, 17], 'juli': [0, 11, 12, 14], '23': [0], '2025': [0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 13, 14, 15, 16, 17, 18, 19]}


## 🧪 Test: Phrase Queries


### 🗣 Instructor Talking Point:
> A phrase query requires the exact sequence of terms (e.g., "machine learning"). To support this, extend the inverted index to store positions, not just docIDs.

### 🔧 Your Task:
- Implement 2 phrase queries.
- Demonstrate that they return the correct documents.


In [24]:
# Placeholder for phrase query implementation
# You may build a position-aware index or use string search within docs after normalization

# Example:
query1 = "Python software"
query2 = "bylaws"

def phrase_query(query, documents):
    results = []
    query_tokens = normalize_tokens(tokenize(query))
    for doc_id, doc in enumerate(documents):
        doc_tokens = normalize_tokens(tokenize(doc))
        # Search for the sequence in the token list
        for i in range(len(doc_tokens) - len(query_tokens) + 1):
            if doc_tokens[i:i+len(query_tokens)] == query_tokens:
                results.append(doc_id)
                break
    return results

# Run phrase queries on the loaded documents
results1 = phrase_query(query1, documents)
results2 = phrase_query(query2, documents)

print(f"Documents containing the phrase '{query1}': {results1}")
print(f"Documents containing the phrase '{query2}': {results2}")

Documents containing the phrase 'Python software': [0, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14, 15]
Documents containing the phrase 'bylaws': [0, 14]


## 🧠 Additional Challenge: Implement Positional Indexes

Implement Positional Indexes, an advanced version of an inverted index that not only stores which documents a term appears in, but also where (at what positions) the term occurs within each document.

Apply it to the documents you used to produce the Inverted Matrix during the Active Learning Workshop. 

Then compare it against a **Term Document Count Matrix** (which you don't have to implement)

And finally, fill in the blanks in the table below.

| Term | What is it? | How is it used? | Pros | Cons | 
|----------------|-------------|-------------|-------------|-------------|
| Term Frequency (TF)  |  |  |  |
| Inverse Document Frequency (idF weight)  |  |  |  |

Then use the table to prepare talking points on:
- Which implementation you prefer to use for searching bigrams (a.k.a., biwords), pairs of consecutive words in a document, and why?

#### Sample solution:

In [25]:
# Implement Positional Indexes: Store term positions for each document
from collections import defaultdict

def build_positional_index(documents):
    positional_index = defaultdict(lambda: defaultdict(list))
    for doc_id, text in enumerate(documents):
        tokens = normalize_tokens(tokenize(text))
        for pos, token in enumerate(tokens):
            positional_index[token][doc_id].append(pos)
    return positional_index

# Build positional index for the loaded documents
positional_index = build_positional_index(documents)

# Example: Show positions for a sample term
sample_term = 'python'  # Change to any term you want to inspect
print(f"Positions for term '{sample_term}':")
for doc_id, positions in positional_index[sample_term].items():
    print(f"  Document {doc_id}: {positions}")

Positions for term 'python':
  Document 0: [1, 15, 181]
  Document 1: [0, 12, 204, 220, 661, 909]
  Document 2: [0, 9, 32, 38, 61, 110, 145, 149, 158, 168, 185, 195, 223, 238, 258, 365, 437, 474, 694]
  Document 3: [0, 5, 12, 20, 49, 56, 69, 109, 120, 124, 134, 152, 189, 226, 228, 265, 281, 283, 295, 313, 330, 337, 390]
  Document 4: [0, 13, 24, 30, 40, 51, 55, 79, 86, 97, 106, 109, 114, 154, 170, 193, 196, 225, 242, 258, 304, 314, 348]
  Document 5: [0, 4, 21, 26, 41, 50, 57, 83, 106, 217, 277, 361, 374, 385, 465, 475]
  Document 6: [0, 45, 52, 86, 210, 242, 249, 276, 281, 284, 289, 305, 399, 406, 417, 420, 426, 616, 771, 794, 826, 836, 845, 862, 905, 957, 973, 989, 998, 1014, 1038, 1063, 1132, 1195, 1217, 1312, 1315, 1320, 1353, 1364, 1378]
  Document 7: [0, 48, 57, 83, 87, 98, 121, 137, 190, 204, 275, 376, 396]
  Document 8: [1, 266, 1897, 1995, 2001, 2006, 2009, 2015, 2019, 2030, 2103, 2156, 2162, 2374]
  Document 9: [8, 41]
  Document 10: [0, 8, 22, 56, 205, 211, 220, 227, 235, 24

### 📊 Comparison: Positional Index vs. Term Document Count Matrix

A **Positional Index** stores not only which documents a term appears in, but also the exact positions of each term within those documents. In contrast, a **Term Document Count Matrix** (TDCM) simply records the frequency of each term in each document, without any positional information.

| Feature                      | Positional Index                | Term Document Count Matrix (TDCM) |
|------------------------------|---------------------------------|-----------------------------------|
| Stores term positions        | Yes                             | No                                |
| Supports phrase/bigram search| Yes                             | No                                |
| Memory usage                 | Higher                          | Lower                             |
| Query speed for phrases      | Fast                            | Slow (requires scanning)          |
| Useful for                   | Phrase queries, proximity search | Keyword frequency analysis         |
| Implementation complexity    | Higher                          | Lower                             |

**Summary:**
- Use a positional index for advanced search features like phrase and proximity queries.
- Use a TDCM for simple keyword frequency analysis and ranking.

### 📝 Completed Table: Term Frequency (TF) and Inverse Document Frequency (IDF weight)

| Term                           | What is it?                                                                 | How is it used?                          | Pros                                   | Cons                                   |
|-------------------------------|----------------------------------------------------------------------------|------------------------------------------|----------------------------------------|----------------------------------------|
| Term Frequency (TF)            | Number of times a term appears in a document                               | Measures term importance in a document   | Simple, fast, good for ranking         | Doesn't account for common terms       |
| Inverse Document Frequency     | Log ratio of total docs to docs containing the term                        | Weights rare terms higher                | Reduces impact of common words         | Can overweight rare, irrelevant terms  |
| (IDF weight)                   | Used with TF to score term relevance: TF-IDF = TF * IDF                    | Improves search relevance                | Highlights unique, important terms     | Requires global corpus statistics      |

Let's implement the Inverse Document Frequency (IDF)

In [26]:
# Calculate Inverse Document Frequency (IDF) for each term using the positional index
import math

def compute_idf_from_positional_index(positional_index, total_docs):
    idf_scores = {}
    for term, doc_positions in positional_index.items():
        doc_count = len(doc_positions)
        if doc_count > 0:
            # --- IGNORE ---
            # print(f"Computing IDF for {term} '{total_docs}' / {doc_count}")
            idf_scores[term] = math.log(total_docs / doc_count)
        else:
            idf_scores[term] = 0.0
    return idf_scores

# Compute IDF for all terms using the positional index
idf_scores = compute_idf_from_positional_index(positional_index, len(documents))

# Example: Show IDF for a sample term
# NOTE : We applied stemming in the normalization step, so the term may not match exactly if it was stemmed
sample_term = 'softwar'  # Change to any term you want to inspect
print(f"IDF for term '{sample_term}': {idf_scores.get(sample_term, 0.0)}")

IDF for term 'softwar': 0.28768207245178085


### 📐 Mathematical Explanation: Inverse Document Frequency (IDF)

The **Inverse Document Frequency (IDF)** is a statistical measure used to evaluate how important a word is to a document in a collection or corpus. The intuition is that terms that appear in many documents are less informative than those that appear in few.

The IDF for a term $t$ is defined as:

$$
\text{IDF}(t) = \log\left(\frac{N}{n_t}\right)
$$

Where:
- $N$ is the total number of documents in the corpus.
- $n_t$ is the number of documents containing the term $t$.

---

#### Example Calculation

Suppose:
- $N = 20$ (total documents)
- $n_{\text{softwar}} = 15$ (documents containing the term 'softwar')

Then:

$$
\text{IDF}(\text{softwar}) = \log\left(\frac{20}{15}\right) \approx 0.2877
$$

So, the IDF for term 'softwar' is $0.2877$.

A higher IDF value means the term is rare across the corpus, while a lower value means the term is common. IDF is often used in combination with Term Frequency (TF) to compute the TF-IDF score:

$$
\text{TF-IDF}(t, d) = \text{TF}(t, d) \times \text{IDF}(t)
$$

Where $\text{TF}(t, d)$ is the frequency of term $t$ in document $d$.

This weighting helps highlight terms that are both frequent in a document and rare across the corpus, improving search relevance and document ranking.

In [27]:
# Calculate Term Frequency (TF) for the term 'softwar' in each document
# TF is simply the number of times the term appears in a document
import pandas as pd

term = 'softwar'  # Use the stemmed term as before

tf_values = []
for doc_id, doc in enumerate(documents):
    tokens = normalize_tokens(tokenize(doc))
    tf = tokens.count(term)
    tf_values.append(tf)
    print(f"TF for term '{term}' in Document {doc_id}: {tf}")

# Display as a DataFrame for clarity
TF_df = pd.DataFrame({'Document': list(range(len(documents))), 'TF': tf_values})
print(TF_df)

TF for term 'softwar' in Document 0: 3
TF for term 'softwar' in Document 1: 0
TF for term 'softwar' in Document 2: 1
TF for term 'softwar' in Document 3: 1
TF for term 'softwar' in Document 4: 2
TF for term 'softwar' in Document 5: 1
TF for term 'softwar' in Document 6: 3
TF for term 'softwar' in Document 7: 1
TF for term 'softwar' in Document 8: 1
TF for term 'softwar' in Document 9: 1
TF for term 'softwar' in Document 10: 4
TF for term 'softwar' in Document 11: 1
TF for term 'softwar' in Document 12: 1
TF for term 'softwar' in Document 13: 1
TF for term 'softwar' in Document 14: 3
TF for term 'softwar' in Document 15: 1
TF for term 'softwar' in Document 16: 0
TF for term 'softwar' in Document 17: 0
TF for term 'softwar' in Document 18: 0
TF for term 'softwar' in Document 19: 0
    Document  TF
0          0   3
1          1   0
2          2   1
3          3   1
4          4   2
5          5   1
6          6   3
7          7   1
8          8   1
9          9   1
10        10   4
11    

#### 📊 Example: Term Frequency (TF) and TF-IDF Calculation for 'softwar' in Document 0

TF for term 'softwar' in Document 0:

$$
\text{TF}(\text{softwar}, 0) = n_{\text{softwar}, 0}
$$

TF-IDF for term 'softwar' in Document 0:

$$
\text{TF-IDF}(\text{softwar}, 0) = \text{TF}(\text{softwar}, 0) \times \text{IDF}(\text{softwar}) = n_{\text{softwar}, 0} \times \text{IDF}(\text{softwar})
$$

Where:
- $n_{\text{softwar}, 0}$ is the number of times 'softwar' appears in Document 0.
- $\text{IDF}(\text{softwar})$ is the previously calculated IDF value for 'softwar'.

Substitute the actual values from your output to see the final TF-IDF score for Document 0.

## 📊 Corpus & Search Analytics with Visualizations
### Task completed by Muthu

Now that we have built the full pipeline — tokenizer, normalizer, inverted index, positional index, and TF-IDF scores — let's **visualize** what our data actually looks like. Charts and graphs help us spot patterns that are hard to see in raw numbers.

Think of it like this: reading a table of 2000 numbers is like trying to memorize a phone book. But a single bar chart can tell you the whole story in one glance.

**What You'll See:**
1. **Corpus Overview** — How big is each document? How rich is our vocabulary?
2. **Zipf's Law** — Which words appear the most, and does our corpus follow a famous pattern?
3. **TF-IDF Heatmap** — Which words matter most in which documents?
4. **Boolean vs. Ranked Search** — How does plain keyword matching compare to TF-IDF ranked results?

In [ ]:
# ============================================================
# Task completed by Muthu — Corpus Overview Visualization
# ============================================================
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Use a clean, readable style for all plots
sns.set_style('whitegrid')

# --- Compute corpus-level statistics ---
# Tokenize and normalize every document so we can measure sizes
all_raw_tokens = [tokenize(doc) for doc in documents]
all_norm_tokens = [normalize_tokens(t) for t in all_raw_tokens]

doc_lengths_raw = [len(t) for t in all_raw_tokens]       # words before cleanup
doc_lengths_norm = [len(t) for t in all_norm_tokens]      # words after cleanup
unique_vocab = set(token for tokens in all_norm_tokens for token in tokens)

print(f"Total documents        : {len(documents)}")
print(f"Total raw tokens       : {sum(doc_lengths_raw)}")
print(f"Total normalized tokens: {sum(doc_lengths_norm)}")
print(f"Unique vocabulary size : {len(unique_vocab)}")

# --- Bar chart: Document lengths (raw vs normalized) ---
fig, ax = plt.subplots(figsize=(12, 5))
x = np.arange(len(documents))
width = 0.35

bars1 = ax.bar(x - width/2, doc_lengths_raw, width, label='Raw tokens', color='steelblue')
bars2 = ax.bar(x + width/2, doc_lengths_norm, width, label='After normalization', color='coral')

ax.set_xlabel('Document ID', fontsize=12)
ax.set_ylabel('Number of Tokens', fontsize=12)
ax.set_title('Document Lengths: Raw vs. Normalized Token Counts', fontsize=14)
ax.set_xticks(x)
ax.legend()
plt.tight_layout()
plt.show()

**What do we see here?**

- Each document shrinks significantly after normalization — stop words like "the", "is", "and" are removed, and stemming collapses different word forms into one.
- Some documents are much longer than others. Longer documents will naturally have more terms in the inverted index, which can influence search results.
- The **vocabulary size** tells us how "rich" our corpus is. If it exceeds 2,000 unique words, we have a solid foundation for meaningful search.

> Think of normalization like packing for a trip — you remove the stuff you don't need (stop words) and fold similar items together (stemming), so everything fits better and you can find what you need faster.

In [ ]:
# ============================================================
# Task completed by Muthu — Zipf's Law: Term Frequency Distribution
# ============================================================
# Zipf's Law says: a few words appear VERY often, and most words
# appear rarely. It's like social media followers — a handful of
# accounts have millions, but most people have a few hundred.

from collections import Counter

# Count how often each normalized term appears across the entire corpus
all_terms = [token for tokens in all_norm_tokens for token in tokens]
term_counts = Counter(all_terms)

# Sort by frequency (most common first)
sorted_terms = term_counts.most_common()
ranks = range(1, len(sorted_terms) + 1)
frequencies = [count for _, count in sorted_terms]

# --- Plot 1: Top 30 most frequent terms (bar chart) ---
fig, axes = plt.subplots(1, 2, figsize=(18, 6))

top_n = 30
top_terms = sorted_terms[:top_n]
axes[0].barh([t for t, _ in top_terms][::-1], [c for _, c in top_terms][::-1], color='teal')
axes[0].set_xlabel('Frequency', fontsize=12)
axes[0].set_title(f'Top {top_n} Most Frequent Terms (After Normalization)', fontsize=13)

# --- Plot 2: Zipf's Law (log-log rank vs frequency) ---
axes[1].loglog(ranks, frequencies, marker='.', linestyle='none', color='darkorange', alpha=0.6)
axes[1].set_xlabel('Rank (log scale)', fontsize=12)
axes[1].set_ylabel('Frequency (log scale)', fontsize=12)
axes[1].set_title("Zipf's Law: Rank vs. Frequency (Log-Log)", fontsize=13)

plt.tight_layout()
plt.show()

print(f"\nMost common term: '{sorted_terms[0][0]}' appears {sorted_terms[0][1]} times")
print(f"Least common terms appear only {sorted_terms[-1][1]} time(s)")
print(f"Total unique terms: {len(sorted_terms)}")

**Key Insight — Zipf's Law in Action**

- The **left chart** shows the most common words in our corpus. Terms like "python", "p", and "board" dominate because our data comes from Python Software Foundation blog posts.
- The **right chart** plots every term's rank against its frequency on a log-log scale. If the dots form a roughly straight line sloping downward, our corpus follows **Zipf's Law** — a well-known pattern in natural language where a small number of words do most of the heavy lifting.
- **Why does this matter for search?** The most frequent terms (like "python") appear in almost every document, so they are *not* very useful for distinguishing one document from another. This is exactly why **IDF** (Inverse Document Frequency) is important — it gives higher weight to rare, distinctive terms.

> Analogy: If every student in your class owns a backpack, "owns a backpack" tells you nothing about who is who. But "plays the tuba" — that narrows it down fast. IDF works the same way.

In [ ]:
# ============================================================
# Task completed by Muthu — TF-IDF Heatmap: Top Terms x Documents
# ============================================================
# A heatmap is like a color-coded grade sheet: rows are terms,
# columns are documents, and brighter colors mean higher TF-IDF
# scores — i.e., that term is especially important in that document.

# Step 1: Pick the top 15 terms by IDF (most distinctive terms)
top_idf_terms = sorted(idf_scores.items(), key=lambda x: x[1], reverse=True)[:15]
top_terms_list = [term for term, _ in top_idf_terms]

# Step 2: Build a TF-IDF matrix (rows=terms, columns=documents)
tfidf_matrix = []
for term in top_terms_list:
    row = []
    for doc_id in range(len(documents)):
        # TF = number of times this term appears in this document
        tf = all_norm_tokens[doc_id].count(term)
        # IDF = already computed
        idf = idf_scores.get(term, 0.0)
        row.append(tf * idf)
    tfidf_matrix.append(row)

# Step 3: Create a DataFrame and plot the heatmap
tfidf_df = pd.DataFrame(
    tfidf_matrix,
    index=top_terms_list,
    columns=[f'Doc {i}' for i in range(len(documents))]
)

fig, ax = plt.subplots(figsize=(16, 7))
sns.heatmap(
    tfidf_df,
    annot=True,          # show the numeric values inside each cell
    fmt='.2f',           # round to 2 decimal places
    cmap='YlOrRd',       # yellow-to-red color gradient
    linewidths=0.5,
    ax=ax
)
ax.set_title('TF-IDF Heatmap: Top 15 Distinctive Terms Across Documents', fontsize=14)
ax.set_xlabel('Document', fontsize=12)
ax.set_ylabel('Term (stemmed)', fontsize=12)
plt.tight_layout()
plt.show()

print("Brighter/darker red cells = higher TF-IDF = term is especially relevant to that document.")

**What do we see here?**

- Most cells are **pale yellow (near zero)** — this means most distinctive terms only appear in one or two documents. That's exactly what makes them distinctive!
- The **bright red hotspots** tell us which document "owns" a particular topic. For example, if the term "bylaw" lights up only in Documents 0 and 14, those are likely the documents about PSF governance changes.
- Terms that are uniformly zero across most documents have **high IDF** — they are rare. When they *do* appear, they strongly signal what that document is about.

> Think of it like a class seating chart with colored stickers: if a sticker only appears on one desk, you immediately know whose it is. That's high TF-IDF in action.

In [ ]:
# ============================================================
# Task completed by Muthu — Boolean vs. TF-IDF Ranked Search
# ============================================================
# Boolean search just says "yes/no, this document matches."
# TF-IDF ranked search says "this document matches, and HERE'S
# HOW RELEVANT it is." Let's compare them side by side.

def boolean_search(query, inverted_index):
    """Return document IDs that contain ALL query terms (AND logic)."""
    query_terms = normalize_tokens(tokenize(query))
    if not query_terms:
        return []
    # Start with docs matching the first term, then intersect
    result_set = set(inverted_index.get(query_terms[0], []))
    for term in query_terms[1:]:
        result_set &= set(inverted_index.get(term, []))
    return sorted(result_set)

def ranked_search(query, documents, idf_scores):
    """Return documents ranked by total TF-IDF score for the query."""
    query_terms = normalize_tokens(tokenize(query))
    scores = {}
    for doc_id in range(len(documents)):
        doc_tokens = all_norm_tokens[doc_id]
        total_score = 0.0
        for term in query_terms:
            tf = doc_tokens.count(term)
            idf = idf_scores.get(term, 0.0)
            total_score += tf * idf
        if total_score > 0:
            scores[doc_id] = round(total_score, 4)
    # Sort by score descending
    return dict(sorted(scores.items(), key=lambda x: x[1], reverse=True))

# --- Run both search methods on the same query ---
search_query = "python community"
print(f"Search query: '{search_query}'\n")

boolean_results = boolean_search(search_query, inverted_index)
ranked_results = ranked_search(search_query, documents, idf_scores)

print(f"Boolean Search (AND): {len(boolean_results)} documents match → {boolean_results}")
print(f"Ranked Search (TF-IDF): {len(ranked_results)} documents, sorted by relevance:")
for doc_id, score in ranked_results.items():
    print(f"  Doc {doc_id}: TF-IDF score = {score}")

# --- Side-by-side bar chart ---
fig, axes = plt.subplots(1, 2, figsize=(16, 5))

# Left: Boolean results (binary — match or no match)
boolean_flags = [1 if i in boolean_results else 0 for i in range(len(documents))]
colors_bool = ['seagreen' if f == 1 else 'lightgray' for f in boolean_flags]
axes[0].bar(range(len(documents)), boolean_flags, color=colors_bool)
axes[0].set_xlabel('Document ID', fontsize=12)
axes[0].set_ylabel('Match (1=Yes, 0=No)', fontsize=12)
axes[0].set_title(f'Boolean AND Search: "{search_query}"', fontsize=13)
axes[0].set_xticks(range(len(documents)))
axes[0].set_yticks([0, 1])

# Right: Ranked results (continuous scores)
ranked_scores_all = [ranked_results.get(i, 0.0) for i in range(len(documents))]
colors_rank = ['darkorange' if s > 0 else 'lightgray' for s in ranked_scores_all]
axes[1].bar(range(len(documents)), ranked_scores_all, color=colors_rank)
axes[1].set_xlabel('Document ID', fontsize=12)
axes[1].set_ylabel('TF-IDF Relevance Score', fontsize=12)
axes[1].set_title(f'TF-IDF Ranked Search: "{search_query}"', fontsize=13)
axes[1].set_xticks(range(len(documents)))

plt.tight_layout()
plt.show()

**What do we see here?**

- **Boolean search (left)** is all-or-nothing: a document either contains *all* query terms or it doesn't. Every matching document looks the same — there's no ranking.
- **TF-IDF ranked search (right)** shows a gradient: some documents are *much more* relevant than others. The tallest orange bars are the best matches — they mention the query terms frequently *and* those terms are distinctive.
- Notice how ranked search can surface the same documents as Boolean, but it also tells you **which one to read first**. That's the real power of TF-IDF.

> Analogy: Boolean search is like asking "who in this room has read Harry Potter?" — everyone raises their hand. Ranked search is like asking "who can quote the most passages?" — now you know who the real fans are.

### Key Takeaways — Task completed by Muthu

1. **Normalization matters.** Removing stop words and stemming reduced token counts significantly, making the inverted index leaner and search more accurate.
2. **Zipf's Law is real.** A tiny fraction of words account for most of the text. This is why IDF exists — to downweight those ultra-common terms.
3. **TF-IDF heatmaps reveal document identity.** Each document has a unique "fingerprint" of high-scoring terms that distinguish it from the rest of the corpus.
4. **Ranked search beats Boolean for usability.** Boolean tells you *what matches*; TF-IDF tells you *what matters most*. Real search engines (Google, Bing) all use ranked retrieval for this reason.

| Visualization | What It Answers |
|---|---|
| Document Length Bar Chart | How big are our documents? How much does normalization trim? |
| Zipf's Law Log-Log Plot | Does our corpus follow natural language frequency patterns? |
| TF-IDF Heatmap | Which terms are most important to which documents? |
| Boolean vs. Ranked Comparison | Why is ranked search better than yes/no matching? |

## 🧠 Reflection

### What We Learned

This workshop walked us through the **complete Information Retrieval pipeline** — from raw text documents all the way to ranked search results. Here are the key reflections:

1. **Tokenization and normalization are the unsung heroes.** Before we can search anything, we need to break text into tokens and clean them up. Removing stop words and applying stemming dramatically reduced noise and made our inverted index leaner and more focused.

2. **The inverted index is elegantly simple.** At its core, it's just a dictionary mapping words to document IDs — yet this simple structure powers the search functionality behind major search engines. Adding positional information extends it to support phrase queries without much extra complexity.

3. **TF-IDF bridges the gap between matching and ranking.** Boolean search tells us *which* documents match, but TF-IDF tells us *how well* they match. The visualizations (especially the Boolean vs. Ranked comparison chart) made this distinction crystal clear — ranked retrieval surfaces the most relevant documents first.

4. **Zipf's Law is real and visible in our data.** A small number of terms (like "python") appeared in nearly every document, while thousands of terms appeared only once or twice. This validates why IDF weighting is essential — without it, common terms would dominate every query.

5. **Visualizations turn abstract numbers into understanding.** The heatmap, bar charts, and Zipf's Law plot transformed raw TF-IDF scores and token counts into intuitive, visual stories that are far easier to interpret than tables of numbers.

### Limitations
- Our corpus is small (20 blog posts from a single source), which limits vocabulary diversity.
- The RSS feed may return different content over time, making results non-reproducible without caching.
- We used simple stemming (Porter Stemmer) — lemmatization or more advanced NLP could improve normalization quality.

### What Could Be Improved
- Use a larger, more diverse corpus (e.g., Wikipedia articles across multiple topics).
- Implement cosine similarity for multi-term query ranking instead of summing TF-IDF scores.
- Add a simple user interface to make querying interactive.

In [ ]:
# Query processor: given a query string and an inverted index,
# return a list of all document IDs that contain any of the query terms (Boolean OR).
def query_processor(query, inverted_index):
    results = []
    for term in query.split():
        if term in inverted_index:
            results.extend(inverted_index[term])
    return list(set(results))  # return unique doc IDs

## 🧠 Push your code to the course GitHub repository

Update your team number and then run the code. 

In [ ]:
!python utils/submit_assignment.py --notebook IR_InvertedMatrix_Workshop.ipynb --student_id "teamA"